### PR TITLE
enterprise portal flow

### PR DESCRIPTION
**Pull request by kashifch**
@raeeschachar @brittneyexline @zubair-arbi 

Please review

### CODE REVIEW

**Comment by kashifch** [comment]
@brittneyexline @zubair-arbi 

I have removed the idp flow as I found existing tests for it in platform

**Comment by kashifch** [comment]
@raeeschachar 

I have changed some part of LmsApi so that we can do logout using api as well

File: regression/pages/enterprise/enterprise_portal_course_page.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        self.browser.switch_to_frame(iframe)
+
+
+class EnterprisePortalCourseStart(EnterprisePortalCourse):
+    """
+    This class handles the Enterprise portal course start/continue objects
+    """
+    def start_or_continue_course(self):
+        """
+        Click on start/contonue course button
```
Comment (zubair-arbi) [inline regression/pages/enterprise/enterprise_portal_course_page.py]: typo: `continue`

File: regression/pages/enterprise/enterprise_course_enrollment_page.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    def is_browser_on_page(self):
+        """
+        Verifies if the Course confirmation title is visible on the page
+        """
+        return self.q(css='.course-confirmation-title').visible
+
+    def get_course_title(self):
+        """
+        Returns course title
+        Returns:
```
Comment (zubair-arbi) [inline regression/pages/enterprise/enterprise_course_enrollment_page.py]: please remove this `Returns` as we have already mentioned it returns course title

File: regression/pages/enterprise/enterprise_portal_home_page.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    def open_courses_popup(self):
+        """
+        Open courses popup
+        """
+        self.q(css='#__tile2').click()
+        self.wait_for_element_visibility('#__dialog2', 'wait for pop up')
+
+    def fetch_courses_list(self):
+        """
+        Fetch course list
```
Comment (zubair-arbi) [inline regression/pages/enterprise/enterprise_portal_home_page.py]:
```python
"""
Fetch course titles list text.
"""
```

File: regression/pages/enterprise/enterprise_portal_home_page.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        return "Home" in self.q(css='#__xbutton1').attrs('title')[0]
+
+    def open_courses_popup(self):
+        """
+        Open courses popup
+        """
+        self.q(css='#__tile2').click()
+        self.wait_for_element_visibility('#__dialog2', 'wait for pop up')
+
+    def fetch_courses_list(self):
```
Comment (zubair-arbi) [inline regression/pages/enterprise/enterprise_portal_home_page.py]: may we can rename this method to `fetch_courses_title_list`

File: regression/tests/enterprise/test_enterprise_login.py [hunk 5]
```diff
@@ -118,4 +134,54 @@ def login_and_unlink_account(self):
...
+            self.ENTERPRISE_PORTAL_PASSWORD)
+        self.enterprise_portal_home.wait_for_page()
+
+    def access_course(self):
+        """
+        Access the course from portal
+        """
+        # Open the course pop up and look for the desired course
+        self.enterprise_portal_home.open_courses_popup()
+        courses_list = self.enterprise_portal_home.fetch_courses_list()
```
Comment (zubair-arbi) [inline regression/tests/enterprise/test_enterprise_login.py]:
```python
courses_title_list = self.enterprise_portal_home.fetch_courses_title_list()
```

File: regression/tests/enterprise/test_enterprise_login.py [hunk 6]
```diff
@@ -118,4 +134,54 @@ def login_and_unlink_account(self):
...
+        self.assert_(self.ENTERPRISE_COURSE_TITLE in x for x in courses_list)
```
Comment (zubair-arbi) [inline regression/tests/enterprise/test_enterprise_login.py]:
```python
self.assert_(self.ENTERPRISE_COURSE_TITLE in course_title for course_title in courses_title_list)
```

**Comment by zubair-arbi** [comment]
@edx/testeng could some from test engineering team please review these changes?

**Comment by kashifch** [comment]
@brittneyexline 

I have incorporated the feedback, can you also take a look and let me know your feedback from business logic perspective

@raeeschachar @jzoldak 

Can I get one review from the test engineering team as well

File: regression/pages/enterprise/enterprise_portal_login_page.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    def is_browser_on_page(self):
+        """
+        Verifies if the browser is on the correct page
+        """
+        return ENTERPRISE_NAME in self.browser.title
+
+    def login_to_portal(self, username, password):
+        """
+        Login IDP user
+        Args:
```
Comment (zubair-arbi) [inline regression/pages/enterprise/enterprise_portal_login_page.py]: Please use full word `Arguments`: http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python_guidelines.html#docstrings
```python
"""
Login user on IDP portal.

Arguments:
    username (str): Unique identifier for user
    password (str): Password for the provided user
"""
```

File: regression/tests/enterprise/test_enterprise_login.py [hunk 8]
```diff
@@ -4,102 +4,118 @@
...
+        self.enterprise_portal_course_start = \
+            EnterprisePortalCourseStart(self.browser)
+        self.enterprise_portal_course_structure = \
+            EnterprisePortalCourseStructure(self.browser)
+        self.enterprise_course_enrollment = \
+            EnterpriseCourseEnrollment(self.browser)
         self.lms_login = LmsLogin(self.browser)
-        self.idp_login = IDPLogin(self.browser)
-        self.enterprise_logistration = EnterpriseLogistration(self.browser)
+        self.enterprise_edx_logistration = \
```
Comment (zubair-arbi) [inline regression/tests/enterprise/test_enterprise_login.py]: Please keep the assignment in single line as the length of line is not more than 120.

File: regression/pages/enterprise/enterprise_portal_course_page.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+class EnterprisePortalCourseStructure(EnterprisePortalCourse):
+    """
+    This class handles the Enterprise portal course structure objects
+    """
+
+    def open_course_on_edx(self):
+        """
+        Open the edx course page in new browser window
+        Returns:
```
Comment (zubair-arbi) [inline regression/pages/enterprise/enterprise_portal_course_page.py]: Please remove this `Returns` statement as its empty.

File: regression/pages/enterprise/enterprise_portal_home_page.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        Returns:
+                Course titles list:
+        """
+        self.wait_for_course_list()
+        return self.q(css=self.COURSE_LIST_CSS).text
+
+    def open_enterprise_course_page(self, course_title_part):
+        """
+        Go to target enterprise course
+        Args:
```
Comment (zubair-arbi) [inline regression/pages/enterprise/enterprise_portal_home_page.py]: full world `Arguments` with a newline between docstring and `Aguments` for readability.

File: regression/pages/enterprise/enterprise_portal_home_page.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        Open courses popup
+        """
+        self.q(css='#__tile2').click()
+        self.wait_for_element_visibility('#__dialog2', 'wait for pop up')
+
+    def fetch_course_titles_list(self):
+        """
+        Fetch course titles list
+        Returns:
+                Course titles list:
```
Comment (zubair-arbi) [inline regression/pages/enterprise/enterprise_portal_home_page.py]: nit: indentation
```
"""
Fetch list of course titles as text.

Returns:
    str: List of course titles
"""
```

File: regression/tests/enterprise/test_enterprise_login.py (hunk 8 above)
Comment (kashifch) [inline regression/tests/enterprise/test_enterprise_login.py]: We are following the protocol of setting line length to 80

Comment (zubair-arbi) [inline regression/tests/enterprise/test_enterprise_login.py]: @kashifch AFAIK the 80 characters line length is for the docstring and 120 characters line length is for the python code. I am not sure if you have internal team rules so its upto you.

File: regression/pages/enterprise/enterprise_portal_home_page.py (hunk 11 above)
Comment (zubair-arbi) [inline regression/pages/enterprise/enterprise_portal_home_page.py]: I guess we missed this comment

File: regression/tests/enterprise/test_enterprise_login.py (hunk 8 above)
Comment (kashifch) [inline regression/tests/enterprise/test_enterprise_login.py]: Well we are following the default pep8 recommendations
https://www.python.org/dev/peps/pep-0008/#maximum-line-length

This does seems a bit extreme to me but as the entire repo is following this pattern so am I

File: regression/tests/enterprise/test_enterprise_login.py [hunk 12]
```diff
@@ -4,102 +4,118 @@
...
 class TestEnterpriseLogin(WebAppTest):
     """
     Test Enterprise Login
     """
-
-    EDX_USER_EMAIL = os.environ.get('USER_LOGIN_EMAIL')
-    EDX_USER_PASSWORD = os.environ.get('USER_LOGIN_PASSWORD')
-    IDP_USERNAME = os.environ.get('IDP_USERNAME')
-    IDP_PASSWORD = os.environ.get('IDP_PASSWORD')
+    ENTERPRISE_PORTAL_USERNAME = os.environ.get('ENTERPRISE_PORTAL_USERNAME')
```
Comment (kashifch) [inline regression/tests/enterprise/test_enterprise_login.py]: I am using TestCourseDataSAP and mtd_learner7, however this information is stored in environment variables so there is no dependency and we can use any user/course on run time

File: regression/pages/enterprise/enterprise_portal_course_page.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
+"""
+Enterprise portal course start page
+"""
+from bok_choy.page_object import PageObject
+
+
+class EnterprisePortalCourse(PageObject):
```
Comment (kashifch) [inline regression/pages/enterprise/enterprise_portal_course_page.py]: I would prefer generic pages with information of different portals stored in a dictionary, however if there are significant differences in flow and css between different portals we might need to create pages for each portal. I am merging this branch as well but will do something about it in next PR

**Comment by kashifch** [comment]
jenkins run microsites

**Pull request merged**